File: Beecrowd_Parte2/Gloup_Computing_1452.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_APPS 101
#define MAX_NAME_LENGTH 21

int main() {
    int N, M, i, j, k;
    char serverApps[200][MAX_APPS][MAX_NAME_LENGTH];
    char clientApps[200][MAX_APPS][MAX_NAME_LENGTH];
    int serverAppCounts[200], clientAppCounts[200];

    while (1) {
        scanf("%d %d", &N, &M);
        if (N == 0 && M == 0) break;

        // Leitura dos servidores
        for (i = 0; i < N; i++) {
            scanf("%d", &serverAppCounts[i]);
            for (j = 0; j < serverAppCounts[i]; j++) {
                scanf("%s", serverApps[i][j]);
            }
        }

        // Leitura dos clientes
        for (i = 0; i < M; i++) {
            scanf("%d", &clientAppCounts[i]);
            for (j = 0; j < clientAppCounts[i]; j++) {
                scanf("%s", clientApps[i][j]);
            }
        }

        // Contagem das conexões únicas
        int totalConnections = 0;
        for (i = 0; i < M; i++) { // para cada cliente
            for (j = 0; j < N; j++) { // para cada servidor
                int isConnected = 0;
                for (k = 0; k < clientAppCounts[i]; k++) {
                    int l;
                    for (l = 0; l < serverAppCounts[j]; l++) {
                        if (strcmp(clientApps[i][k], serverApps[j][l]) == 0) {
                            isConnected = 1;
                            break;
                        }
                    }
                    if (isConnected) break;
                }
                if (isConnected) totalConnections++;
            }
        }

        // Imprime resultado para este caso de teste
        printf("%d\n", totalConnections);
    }

    return 0;
}

File: Beecrowd_Parte2/Gloup_Computing_1452.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Gloup_Computing_1452 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        while (true) {
            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());

            if (N == 0 && M == 0) {
                break;
            }

            // Leitura dos servidores
            Set<String>[] serverApps = new HashSet[N];
            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                int Qi = Integer.parseInt(st.nextToken());
                serverApps[i] = new HashSet<>();
                for (int j = 0; j < Qi; j++) {
                    serverApps[i].add(st.nextToken());
                }
            }

            // Leitura dos clientes
            Set<String>[] clientApps = new HashSet[M];
            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                int Pj = Integer.parseInt(st.nextToken());
                clientApps[i] = new HashSet<>();
                for (int j = 0; j < Pj; j++) {
                    clientApps[i].add(st.nextToken());
                }
            }

            // Contagem das conexões únicas
            int totalConnections = 0;
            for (int j = 0; j < M; j++) { // para cada cliente
                for (int i = 0; i < N; i++) { // para cada servidor
                    boolean isConnected = false;
                    for (String app : clientApps[j]) {
                        if (serverApps[i].contains(app)) {
                            isConnected = true;
                            break;
                        }
                    }
                    if (isConnected) {
                        totalConnections++;
                    }
                }
            }

            // Imprime resultado para este caso de teste
            System.out.println(totalConnections);
        }

        br.close();
    }
}
